dotCMS/src/main/java/com/dotmarketing/portlets/workflows/model/WorkflowProcessor.java

```diff
@@ -86,41 +86,47 @@ public WorkflowProcessor(Contentlet contentlet, User firingUser) {
 		this.contentlet = contentlet;
 
 		try {
+
 			this.user = firingUser;
+
 			WorkflowStep contentStep = getWorkflowAPI().findStepByContentlet(contentlet);
-			if(null != contentStep) {
+			if (null != contentStep) {
 				scheme = getWorkflowAPI().findScheme(contentStep.getSchemeId());
 			}
 			task = getWorkflowAPI().findTaskByContentlet(contentlet);
 
 			String workflowActionId = contentlet.getStringProperty(Contentlet.WORKFLOW_ACTION_KEY);
-			if (!UtilMethods.isSet(workflowActionId) && task.isNew() && null !=  scheme){
-				workflowActionId=scheme.getEntryActionId();
+			if (UtilMethods.isSet(workflowActionId)) {
+				action = findAction(contentlet, workflowActionId);
 			}
 
-			if (!UtilMethods.isSet(workflowActionId)) {
-				if (null !=  scheme && scheme.isMandatory() ) {
-					throw new DotWorkflowException(LanguageUtil.get(firingUser, "message.workflow.error.mandatory.action.type") + contentlet.getStructure().getName());
-				}
+			//If we found and action and we don't have a workflow we can search for it
+			if (null != action && null == scheme) {
+				scheme = getWorkflowAPI().findScheme(action.getSchemeId());
+			}
 
-				return;
+			if (!UtilMethods.isSet(workflowActionId) && task.isNew() && null != scheme) {
+				workflowActionId = scheme.getEntryActionId();
+				action = findAction(contentlet, workflowActionId);
 			}
 
+			if (!UtilMethods.isSet(workflowActionId)) {
+				if (null != scheme && scheme.isMandatory()) {
+					throw new DotWorkflowException(LanguageUtil
+							.get(user, "message.workflow.error.mandatory.action.type")
+							+ contentlet.getStructure().getName());
+				}
 
-			try{
-				action = getWorkflowAPI().findAction(workflowActionId, user);
-			}
-			catch(Exception ex){
-				throw new DotWorkflowException(LanguageUtil.get(firingUser, "message.workflow.error.invalid.action") + contentlet.getStringProperty(Contentlet.WORKFLOW_ACTION_KEY));
+				return;
 			}
 
-
 			if(action.requiresCheckout()){
-				try{
+				try {
 					APILocator.getContentletAPI().canLock(contentlet, user);
-				}
-				catch(Exception ex){
-					throw new DotWorkflowException(LanguageUtil.get(firingUser, "message.workflow.error.content.requires.lock") + contentlet.getStructure().getName());
+				} catch (Exception ex) {
+					throw new DotWorkflowException(LanguageUtil
+							.get(user, "message.workflow.error.content.requires.lock")
+							+ contentlet.getStructure().getName(), ex);
```

Comment:
![MINOR](https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR') Remove this use of "getStructure"; it is deprecated. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3ACallToDeprecatedMethod)